### ISSUE TITLE
Reproducible performance benchmarks

### ISSUE THREAD

**Issue by florianscheidl** [ecmwf/WeatherGenerator#2612]
### Describe the task. It can be a feature, documentation, etc.

In recent performance experiments, we have seen significant variability in peak memory using the same configuration, similarly for throughput results. 
Therefore we need best practices for assessing performance.

Requirements:
- reproducibility: ensure fixed seeds (especially in performance-relevant components like data sampling).
- representative sample size: to ensure that the variability in hardware and the variability in data is sufficiently averaged out (how many samples are required to attain which level of confidence?)
- (potentially) representative samples: instead of relying on a large sample size capturing the data distribution (especially in case of a fixed seed, we might get unlucky and the sampled data for a given training duration is not representative of the training jobs we are optimizing for), we the data could be chosen to reflect the training distribution.

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

https://gitlab.jsc.fz-juelich.de/hedgedoc/hXo2AzgjTfCa4O0wEdOUgQ?both

### Area

- [x] datasets, data readers, data preparation and transfer
- [x] model
- [ ] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [x] performance

**Comment by clessig** [comment]
For the "heavy" configs we are typically interested in "worst-case" performance when all the streams provide dense input. We can usually ensure this by picking the right time period, e.g. the GEOs are dense in 2023.

**Comment by florianscheidl** [comment]
> For the "heavy" configs we are typically interested in "worst-case" performance when all the streams provide dense input. We can usually ensure this by picking the right time period, e.g. the GEOs are dense in 2023.

Thanks for the context! For memory, this makes perfect sense. Do you also apply this to throughput though?

**Comment by clessig** [comment]
> > For the "heavy" configs we are typically interested in "worst-case" performance when all the streams provide dense input. We can usually ensure this by picking the right time period, e.g. the GEOs are dense in 2023.
> 
> Thanks for the context! For memory, this makes perfect sense. Do you also apply this to throughput though?

For the moment I would also focus on the most heavy workload. We can look later if there is a difference to more unbalanced workloads.

**Comment by florianscheidl** [comment]
Assign me.

**Comment by florianscheidl** [comment]
Update: collected insights regarding the sample size and seeds in the linked [hedgedoc](https://gitlab.jsc.fz-juelich.de/hedgedoc/hXo2AzgjTfCa4O0wEdOUgQ?both). The takeaway is that runs with the same configuration vary considerably in throughput, while memory consumption is stable. We should align on best practices for performance experiments, balancing the need for reasonable estimates with practicability, e.g., running 3-5 repetitions per config for 20 minutes. @javak87, how do you do this in other projects?

Concerning representative samples, I plotted the completeness of different datasets used in our common config (ping me if you're interested in a more detailed interactive version of the plot):

<img width="1190" height="536" alt="Image" src="https://github.com/user-attachments/assets/2086fa3a-8b06-4598-901f-0ea122e2f801" />

We see that the datasets we use are nearly complete from early 2017, with gaps. Completeness refers to the share of non-NaN values for a time slot (computed for 15-minute time slots), with resolution reduced in the plot. Thus, setting the training period to start from mid-2017 should be sufficient for the config_operan_georing_avhrr_forecasting_lowres config. **Note**: the datasets have been updated since the time of generating the plots, so they might be more complete.

So once we have merged:
- https://github.com/ecmwf/WeatherGenerator/pull/2627,
- https://github.com/ecmwf/WeatherGenerator/pull/2628,
- https://github.com/ecmwf/WeatherGenerator/pull/2629,

We will open a new PR with a perf_benchmarking_config that sets the start time for config_operan_georing_avhrr_forecasting_lowres to 2017-06-01T00:00.

The plot was generated on the [exp/visualize-input-data](https://github.com/florianscheidl/WeatherGenerator/tree/exp/visualize-input-data) branch in this [directory](https://github.com/florianscheidl/WeatherGenerator/tree/exp/visualize-input-data/packages/science).

**Comment by javak87** [comment]
Thank you for the deep dive into the reproducibility of the performance results.

To compare performance reliably, it is recommended to run the experiments with different seeds. For example, in MLPerf benchmarks, training should be run 10 times, each time with a different seed, to demonstrate consistent performance.

**Comment by florianscheidl** [comment]
> Thank you for the deep dive into the reproducibility of the performance results.
> 
> To compare performance reliably, it is recommended to run the experiments with different seeds. For example, in MLPerf benchmarks, training should be run 10 times, each time with a different seed, to demonstrate consistent performance.

Thanks for your input. What practice do you think we should follow when, e.g., developing and benchmarking a new feature? 10 runs might be a lot; perhaps 5 can give us enough confidence. A related issue is accuracy: by default, we log metrics every 20 batches. If we start logging them more often, we will force more writes and synchronizations, slowing down the code. Can we extract the number of processed samples in a less invasive way? Could we, e.g., do non-blocking writes per device, and aggregate the data afterwards?

**Comment by javak87** [comment]
I agree that 10 runs is too many. As you suggested, five runs may be sufficient. I think we should stick to realistic training scenarios.

We need to collect training metrics, but the evaluation frequency should reflect actual user behavior. For example, evaluating every 100 batches may not be representative, since users usually want to check the score neither too frequently nor too rarely. Repeating this process for every device may also be excessive, especially because we also need to verify that the training loss is decreasing.

Regarding AI performance analysis, one option is to measure the time required to reach a certain accuracy. Another is to focus on throughput per iteration or per epoch. Since the scientific analysis has already been completed, I think it would be more straightforward to focus on throughput and then validate that the loss decreases as expected.

As you also mentioned, we can run the experiments with several seeds to ensure an apples-to-apples comparison.
